lib/dice/src/cert.rs

```diff
@@ -0,0 +1,346 @@
+    }
+
+    const ISSUER_SN_START: usize;
+    const ISSUER_SN_END: usize;
+
+    fn get_issuer_sn(&self) -> SerialNumber {
+        SerialNumber::from_bytes(
+            &self.as_bytes()[Self::ISSUER_SN_START..Self::ISSUER_SN_END]
+                .try_into()
+                .unwrap_lite(),
+        )
+    }
+
+    fn set_issuer_sn(mut self, sn: &SerialNumber) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::ISSUER_SN_START..Self::ISSUER_SN_END]
+            .copy_from_slice(sn.as_bytes());
+
+        self
+    }
+
+    const NOTBEFORE_START: usize;
+    const NOTBEFORE_END: usize;
+
+    fn set_notbefore(mut self, utctime: &[u8; NOTBEFORE_LENGTH]) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::NOTBEFORE_START..Self::NOTBEFORE_END]
+            .copy_from_slice(utctime);
+
+        self
+    }
+
+    const SUBJECT_SN_START: usize;
+    const SUBJECT_SN_END: usize;
+
+    fn get_subject_sn(&self) -> SerialNumber {
+        SerialNumber::from_bytes(
+            &self.as_bytes()[Self::SUBJECT_SN_START..Self::SUBJECT_SN_END]
+                .try_into()
+                .unwrap_lite(),
+        )
+    }
+
+    fn set_subject_sn(mut self, sn: &SerialNumber) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::SUBJECT_SN_START..Self::SUBJECT_SN_END]
+            .copy_from_slice(sn.as_bytes());
+
+        self
+    }
+
+    const PUB_START: usize;
+    const PUB_END: usize;
+
+    fn get_pub(&self) -> &[u8] {
+        &self.as_bytes()[Self::PUB_START..Self::PUB_END]
+    }
+
+    fn set_pub(mut self, pubkey: &[u8; PUBLICKEY_SERIALIZED_LENGTH]) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::PUB_START..Self::PUB_END]
+            .copy_from_slice(pubkey);
+
+        self
+    }
+
+    const SIG_START: usize;
+    const SIG_END: usize;
+
+    fn get_sig(&self) -> &[u8] {
+        &self.as_bytes()[Self::SIG_START..Self::SIG_END]
+    }
+
+    fn set_sig(mut self, sig: &[u8; SIGNATURE_SERIALIZED_LENGTH]) -> Self
+    where
+        Self: Sized,
+    {
+        self.as_mut_bytes()[Self::SIG_START..Self::SIG_END]
+            .copy_from_slice(sig);
+
+        self
+    }
+
+    const SIGNDATA_START: usize;
+    const SIGNDATA_END: usize;
+
+    fn get_signdata(&self) -> &[u8] {
+        &self.as_bytes()[Self::SIGNDATA_START..Self::SIGNDATA_END]
+    }
+
+    fn sign(self, keypair: &Keypair) -> Self
+    where
+        Self: Sized,
+    {
+        let signdata = self.get_signdata();
+        let sig = keypair.sign(signdata);
+
+        self.set_sig(&sig.to_bytes())
+    }
+}
+
+#[derive(Deserialize, Serialize, SerializedSize)]
+pub struct DeviceIdSelfCert(
+    #[serde(with = "BigArray")] [u8; deviceid_cert_tmpl::SIZE],
+);
+
+impl DeviceIdSelfCert {
+    pub fn new(
+        cert_sn: u8,
+        dname_sn: &SerialNumber,
+        keypair: &Keypair,
+    ) -> Self {
+        let mut buf = [0u8; deviceid_cert_tmpl::SIZE];
```

Comment:
```rust
let mut buf = deviceid_cert_tmpl::CERT_TMPL.clone()
```
?

Reply:
This one falls into the "I didn't know I could do that" category. I still qualify as a recovering C programmer I guess :disappointed: